Fix -Wtype-limits warning on mips

aoikonomopoulos opened the pull request:
#### 30f0d48dee99d3348df11e3b0d28e96313b5b631
<pre>
Fix -Wtype-limits warning on mips
<a href="https://bugs.webkit.org/show_bug.cgi?id=248593">https://bugs.webkit.org/show_bug.cgi?id=248593</a>

Reviewed by Mark Lam.

When setupStubCrossArgs is instantiated with NumCrossSources == 0, we
get
    warning: comparison of unsigned expression in ‘&lt; 0’ is always false [-Wtype-limits]

Don&apos;t enter the loop if NumCrossSources is 0. We still need to mark the
parameters unused in the NumCrossSources == 0 path, to avoid
-Wunused-but-set-parameter warnings.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::setupStubCrossArgs):

Canonical link: <a href="https://commits.webkit.org/257409@main">https://commits.webkit.org/257409@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8526782d1362773fab40d3eb9fe38e90c3837e8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98931 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108343 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168598 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8693 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91462 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106314 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104627 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33615 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21511 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89668 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2045 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85475 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1952 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/28812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5100 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6911 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42501 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88330 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3358 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19773 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -198,11 +198,17 @@ class CCallHelpers : public AssemblyHelpers {
-            FPRReg source = sources[i];
+        if constexpr (!NumCrossSources) {
+            (void)destinations;
+            (void)sources;
```
Use `UNUSED_PARAM(destinations)`, etc. instead.

MenloDorian commented on Source/JavaScriptCore/jit/CCallHelpers.h:
```diff
@@ -198,11 +198,17 @@ class CCallHelpers : public AssemblyHelpers {
-        for (unsigned i = 0; i < NumCrossSources; i++) {
-            GPRReg dest = destinations[i];
-            FPRReg source = sources[i];
+        if constexpr (!NumCrossSources) {
```
Why not just `if constexpr (NumCrossSources)`?  Do we really need to handle this do nothing case?

aoikonomopoulos replied:
Not handling the else results in a warning on GCC 9.x (what the 32-bit JSC buildbots use). It's a clean build if we have UNUSED_PARAM at the entry or exit though, just thought that might have been more confusing to readers.

jjgriego replied:
butting in with a suggestion: reformulating the loop in terms of iterators instead of indices might avoid both the UNUSED_PARAM and the type-limits warning 

EDIT ehhh, maybe not actually any cleaner; I spoke too soon, please disregard :P

MenloDorian replied:
There's also `UNUSED_VARIABLE`.

aoikonomopoulos replied:
Right but the warnings are about unused parameters, not variables, so not sure how that's cleaner? One is defined to the other in any case.

MenloDorian reviewed (approved):
r=me, but please check the alternative I proposed before landing.

aoikonomopoulos committed:
Fix -Wtype-limits warning on mips
https://bugs.webkit.org/show_bug.cgi?id=248593

Reviewed by Mark Lam.

When setupStubCrossArgs is instantiated with NumCrossSources == 0, we
get
    warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]

Don't enter the loop if NumCrossSources is 0. We still need to mark the
parameters unused in the NumCrossSources == 0 path, to avoid
-Wunused-but-set-parameter warnings.

* Source/JavaScriptCore/jit/CCallHelpers.h:
(JSC::CCallHelpers::setupStubCrossArgs):

Canonical link: https://commits.webkit.org/257409@main
